modules/apps/101-interchain-swap/keeper/msg_server.go

```diff
@@ -224,63 +262,162 @@ func (k Keeper) OnDepositReceived(ctx sdk.Context, msg *types.MsgDepositRequest)
 	}, nil
 }
 
-func (k Keeper) OndWithdrawReceive(ctx sdk.Context, msg *types.MsgWithdrawRequest) (*types.MsgWithdrawResponse, error) {
+// OnDoubleDepositReceived processes a double deposit request and returns a response or an error.
+func (k Keeper) OnDoubleDepositReceived(ctx sdk.Context, msg *types.MsgDoubleDepositRequest) (*types.MsgDoubleDepositResponse, error) {
 
+	// Validate the message
 	if err := msg.ValidateBasic(); err != nil {
 		return nil, err
 	}
-	pool, found := k.GetInterchainLiquidityPool(ctx, msg.PoolCoin.Denom)
 
+	// // Verify the sender's address
+	secondSenderAcc := k.authKeeper.GetAccount(ctx, sdk.MustAccAddressFromBech32(msg.Senders[1]))
+	senderPrefix, _, err := bech32.Decode(secondSenderAcc.GetAddress().String())
+	if err != nil {
+		return nil, err
+	}
+	if sdk.GetConfig().GetBech32AccountAddrPrefix() != senderPrefix {
+		return nil, errorsmod.Wrapf(types.ErrFailedDoubleDeposit, "first address has to be this chain address (%s)", err)
+	}
+
+	// Retrieve the liquidity pool
+	pool, found := k.GetInterchainLiquidityPool(ctx, msg.PoolId)
 	if !found {
-		return nil, types.ErrNotFoundPool
+		return nil, errorsmod.Wrapf(types.ErrFailedDoubleDeposit, "%s", types.ErrNotFoundPool)
+	}
+
+	// Lock assets from senders to escrow account
+	escrowAccount := types.GetEscrowAddress(pool.EncounterPartyPort, pool.EncounterPartyChannel)
+	// Create a deposit message
+	sendMsg := banktypes.MsgSend{
+		FromAddress: secondSenderAcc.GetAddress().String(),
+		ToAddress:   escrowAccount.String(),
+		Amount:      sdk.NewCoins(*msg.Tokens[1]),
+	}
+
+	deposit := types.EncounterPartyDepositTx{
+		AccountSequence: secondSenderAcc.GetSequence(),
+		Sender:          secondSenderAcc.GetAddress().String(),
+		Token:           msg.Tokens[1],
 	}
 
-	//TODO: need to implement amm part.
-	//feeRate := parms.getPoolFeeRate()
+	rawDepositTx, err := types.ModuleCdc.Marshal(&deposit)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKey := secondSenderAcc.GetPubKey()
+
+	isValid := pubKey.VerifySignature(rawDepositTx, msg.EncounterPartySignature)
+	if !isValid {
+		return nil, errorsmod.Wrapf(types.ErrFailedDoubleDeposit, ":%s", types.ErrInvalidSignature)
+	}
+
+	_, err = k.executeDepositTx(ctx, &sendMsg)
+	if err != nil {
+		return nil, err
+	}
 
-	// calculate output token.
+	// Instantiate an interchain market maker
 	amm := types.NewInterchainMarketMaker(
 		&pool,
 		types.DefaultMaxFeeRate,
 	)
 
+	// Process double asset deposit
+	poolTokens, err := amm.DepositDoubleAsset(msg.Tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	// Increase LP token mint amount
+	for _, token := range poolTokens {
+		pool.AddPoolSupply(*token)
+	}
+
+	// Update pool tokens or switch pool status to 'READY'
+	if pool.Status == types.PoolStatus_POOL_STATUS_READY {
+		for _, token := range msg.Tokens {
+			pool.AddAsset(*token)
+		}
+	} else {
+		pool.Status = types.PoolStatus_POOL_STATUS_READY
+	}
+
+	// Mint voucher tokens for the sender
```

Comment:
Do you split the amount of LP for each asset in the deposit. or just send to the one who initial deposit

Reply:
As I mentioned before, with current formula, it is not possible to get full fund again by token split
currently about initial deposits. I'm allocating  tokenX+tokenY Lp amount. 
In this case, it is possible to full refund his asset.  Please check test demo repo link which I've provided previous time to you

Reply:
Can you put a link here. therefore others can read as well. 

I am no doubt about the total amount of LP.  what I care is where(which chain) the LP tokens your mint. 

In a multi deposit, the number of LPs minted on each chain should correspond to the number of deposits. For example, deposit 100X + 500Y = 600LP into a 50:50 pool, should be 300LP on chainA and 300LP on ChainB, right?

Reply:
--------------- Create Pool ----------------
lp supply amount: 2001000
USDT : 2000000
ETH : 1000
USDT : 2000000
ETH : 1000
----*********----
UserALPAmount: 400200
UserBLPAmount: 1600800
WithdrawUSDT: 22190USDT
WithdrawETH: 19ETH
Worth: 0.002285382568352454
LPprice: 1.142120224064195e-09
DepositAmount: 20000USDT

As you can see, user can't withdraw fully. 
let's check that reason.
At = Bt * (1 - (1 - P_redeemed / P_supply) ** 1/Wt)
if P_redeemed < P_supply, in this case, (1 - (1 - P_redeemed / P_supply) ** 1/Wt) will be less than 1. 
so user can't withdraw his original amount. 
right?
but P_redeemed is same P_supply, in this case, (1 - (1 - P_redeemed / P_supply) ** 1/Wt) will be 1. 
so user can withdraw his amount correctly 
right? 

Please check formula and let me know.

Reply:
so we have to provide P_supply LP amount for initial users. in this case, only possible to withdraw fully